// @anchor-lang/core/dist/esm/idl.js
import { bs58, utf8 } from "./utils/bytes/index.js";
import { inflate, ungzip } from "pako";
import camelCase from "camelcase";
import { Buffer } from "buffer";
import { PublicKey } from "@solana/web3.js";
export function isCompositeAccounts(accountItem) {
    return "accounts" in accountItem;
}
// Account format defined at
// https://github.com/solana-program/program-metadata/blob/734e947d/clients/js/src/generated/accounts/metadata.ts#L123-L138
const PROGRAM_METADATA_PROGRAM_ID = new PublicKey("ProgM6JCCvbYkfKqJYHePx4xxSUSqJp7rh8Lyv7nk7S");
const IDL_METADATA_SEED = "idl";
const ACCOUNT_DISCRIMINATOR_METADATA = 2;
const DATA_SOURCE_DIRECT = 0;
// Only JSON formatted data is currently supported
export const FORMAT_JSON = 1;
const SEED_SIZE = 16;
const DATA_LENGTH_SIZE = 4;
const DATA_LENGTH_PADDING = 5;
const ZEROABLE_OPTION_PUBKEY_SIZE = 32;
const METADATA_HEADER_SIZE = 1 + 32 + ZEROABLE_OPTION_PUBKEY_SIZE + 1 + 1 + SEED_SIZE + 1 + 1 + 1 + 1;
export var MetadataCompression;
(function (MetadataCompression) {
    MetadataCompression[MetadataCompression["None"] = 0] = "None";
    MetadataCompression[MetadataCompression["Gzip"] = 1] = "Gzip";
    MetadataCompression[MetadataCompression["Zlib"] = 2] = "Zlib";
})(MetadataCompression || (MetadataCompression = {}));
export var MetadataEncoding;
(function (MetadataEncoding) {
    MetadataEncoding[MetadataEncoding["None"] = 0] = "None";
    MetadataEncoding[MetadataEncoding["Utf8"] = 1] = "Utf8";
    MetadataEncoding[MetadataEncoding["Base58"] = 2] = "Base58";
    MetadataEncoding[MetadataEncoding["Base64"] = 3] = "Base64";
})(MetadataEncoding || (MetadataEncoding = {}));
function encodeMetadataSeed(seed) {
    const encodedSeed = Buffer.from(utf8.encode(seed));
    if (encodedSeed.length > SEED_SIZE) {
        throw new Error(`Metadata seed '${seed}' exceeds ${SEED_SIZE} bytes`);
    }
    const paddedSeed = Buffer.alloc(SEED_SIZE);
    encodedSeed.copy(paddedSeed);
    return paddedSeed;
}
export function idlAddress(programId) {
    // Canonical metadata uses a null authority seed, which is serialized as `[]`.
    return PublicKey.findProgramAddressSync([
        programId.toBuffer(),
        Buffer.alloc(0),
        encodeMetadataSeed(IDL_METADATA_SEED),
    ], PROGRAM_METADATA_PROGRAM_ID)[0];
}
export function seed() {
    return "idl";
}
export function decodeIdlAccount(data) {
    const minimumSize = METADATA_HEADER_SIZE + DATA_LENGTH_SIZE + DATA_LENGTH_PADDING;
    if (data.length < minimumSize) {
        throw new Error("Metadata account is too small");
    }
    let offset = 0;
    const discriminator = data.readUInt8(offset);
    offset += 1;
    if (discriminator !== ACCOUNT_DISCRIMINATOR_METADATA) {
        throw new Error(`Invalid metadata account discriminator: ${discriminator.toString()}`);
    }
    offset += 32; // program
    offset += ZEROABLE_OPTION_PUBKEY_SIZE; // authority
    offset += 1; // mutable
    offset += 1; // canonical
    offset += SEED_SIZE; // seed
    const encoding = data.readUInt8(offset);
    offset += 1;
    const compression = data.readUInt8(offset);
    offset += 1;
    const format = data.readUInt8(offset);
    if (format !== FORMAT_JSON) {
        throw new Error(`IDL has data format '${format}', only JSON IDLs (${FORMAT_JSON}) are supported`);
    }
    offset += 1;
    const dataSource = data.readUInt8(offset);
    if (dataSource !== DATA_SOURCE_DIRECT) {
        throw new Error(`IDL has source '${dataSource}', only directly embedded data (${DATA_SOURCE_DIRECT}) is supported`);
    }
    offset += 1;
    const dataLength = data.readUInt32LE(offset);
    offset += DATA_LENGTH_SIZE + DATA_LENGTH_PADDING;
    if (data.length < offset + dataLength) {
        throw new Error("Metadata account data is truncated");
    }
    const blob = data.subarray(offset, offset + dataLength);
    const decoded = decodeMetadataData(uncompressMetadataData(blob, compression), encoding);
    return JSON.parse(decoded);
}
export function uncompressMetadataData(data, compression) {
    switch (compression) {
        case MetadataCompression.None:
            return data;
        case MetadataCompression.Gzip:
            return Buffer.from(ungzip(data));
        case MetadataCompression.Zlib:
            return Buffer.from(inflate(data));
        default:
            throw new Error(`Unsupported metadata compression: ${String(compression)}`);
    }
}
export function decodeMetadataData(data, encoding) {
    switch (encoding) {
        // 'None' is actually hex-encoded
        case MetadataEncoding.None:
            return data.toString("hex");
        case MetadataEncoding.Utf8:
            return utf8.decode(data);
        case MetadataEncoding.Base58:
            return bs58.encode(data);
        case MetadataEncoding.Base64:
            return data.toString("base64");
        default:
            throw new Error(`Unsupported metadata encoding: ${String(encoding)}`);
    }
}
/**
 * Convert the given IDL to camelCase.
 *
 * The IDL is generated from Rust which has different conventions compared to
 * JS/TS, e.g. instruction names in Rust are snake_case.
 *
 * The conversion happens automatically for programs, however, if you are using
 * internals such as `BorshInstructionCoder` and you only have the original
 * (not camelCase) IDL, you might need to use this function.
 *
 * @param idl IDL to convert to camelCase
 * @returns camelCase version of the IDL
 */
export function convertIdlToCamelCase(idl) {
    const KEYS_TO_CONVERT = ["name", "path", "account", "relations", "generic"];
    // `my_account.field` is getting converted to `myAccountField` but we
    // need `myAccount.field`.
    const toCamelCase = (s) => s
        .split(".")
        .map((part) => camelCase(part, { locale: false }))
        .join(".");
    const recursivelyConvertNamesToCamelCase = (obj) => {
        for (const key in obj) {
            const val = obj[key];
            if (KEYS_TO_CONVERT.includes(key)) {
                obj[key] = Array.isArray(val) ? val.map(toCamelCase) : toCamelCase(val);
            }
            else if (typeof val === "object") {
                recursivelyConvertNamesToCamelCase(val);
            }
        }
    };
    const camelCasedIdl = structuredClone(idl);
    recursivelyConvertNamesToCamelCase(camelCasedIdl);
    return camelCasedIdl;
}
/** Conveniently handle all defined field kinds with proper type support. */
export function handleDefinedFields(fields, unitCb, namedCb, tupleCb) {
    // Unit
    if (!(fields === null || fields === void 0 ? void 0 : fields.length))
        return unitCb();
    // Named
    if (fields[0].name) {
        return namedCb(fields);
    }
    // Tuple
    return tupleCb(fields);
}
